File: src/comm_debugger/codec.py
from __future__ import annotations

import re
from typing import Literal


PayloadFormat = Literal["text", "hex"]
LineEnding = Literal["none", "cr", "lf", "crlf"]

_LINE_ENDINGS: dict[LineEnding, bytes] = {
    "none": b"",
    "cr": b"\r",
    "lf": b"\n",
    "crlf": b"\r\n",
}


def parse_hex(value: str) -> bytes:
    """Parse common spaced or compact hexadecimal input."""
    cleaned = re.sub(r"0[xX]", "", value)
    cleaned = re.sub(r"[\s,;:_-]+", "", cleaned)

    if not cleaned:
        return b""
    if not re.fullmatch(r"[0-9a-fA-F]+", cleaned):
        raise ValueError("HEX 数据包含非十六进制字符")
    if len(cleaned) % 2:
        raise ValueError("HEX 数据必须由完整字节组成（每字节两位）")
    return bytes.fromhex(cleaned)


def encode_payload(
    value: str,
    payload_format: PayloadFormat,
    text_encoding: str = "utf-8",
    line_ending: LineEnding = "none",
) -> bytes:
    if payload_format == "hex":
        payload = parse_hex(value)
    else:
        try:
            payload = value.encode(text_encoding)
        except LookupError as exc:
            raise ValueError(f"不支持的文本编码：{text_encoding}") from exc
        except UnicodeEncodeError as exc:
            raise ValueError(f"文本无法使用 {text_encoding} 编码") from exc

    return payload + _LINE_ENDINGS[line_ending]


def format_data(data: bytes) -> tuple[str, str]:
    return data.hex(" ").upper(), data.decode("utf-8", errors="replace")

File: server/src/core/codec.ts
import iconv from "iconv-lite";
import type { DataFormat, LineEnding, TextEncoding } from "./types.js";

const lineEndings: Record<LineEnding, Buffer> = {
  none: Buffer.alloc(0),
  cr: Buffer.from("\r"),
  lf: Buffer.from("\n"),
  crlf: Buffer.from("\r\n"),
};

export function parseHex(value: string): Buffer {
  const cleaned = value
    .replaceAll(/0[xX]/g, "")
    .replaceAll(/[\s,;:_-]+/g, "");
  if (!cleaned) return Buffer.alloc(0);
  if (!/^[0-9a-fA-F]+$/.test(cleaned)) {
    throw new Error("HEX 数据包含非十六进制字符");
  }
  if (cleaned.length % 2 !== 0) {
    throw new Error("HEX 数据必须由完整字节组成（每字节两位）");
  }
  return Buffer.from(cleaned, "hex");
}

export function encodePayload(
  value: string,
  format: DataFormat,
  textEncoding: TextEncoding = "utf-8",
  lineEnding: LineEnding = "none",
): Buffer {
  let payload: Buffer;
  if (format === "hex") {
    payload = parseHex(value);
  } else if (textEncoding === "ascii") {
    if ([...value].some((character) => character.codePointAt(0)! > 0x7f)) {
      throw new Error("文本无法使用 ascii 编码");
    }
    payload = Buffer.from(value, "ascii");
  } else if (textEncoding === "gbk") {
    payload = iconv.encode(value, "gbk");
    if (iconv.decode(payload, "gbk") !== value) {
      throw new Error("文本无法使用 gbk 编码");
    }
  } else {
    payload = Buffer.from(value, "utf8");
  }
  return Buffer.concat([payload, lineEndings[lineEnding]]);
}

export function formatData(data: Buffer): { hex: string; text: string } {
  return {
    hex: data.toString("hex").toUpperCase().replaceAll(/(..)(?=.)/g, "$1 "),
    text: data.toString("utf8"),
  };
}
